regression detector: change baseline variant from latest main to merge base

goxberry opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

This PR changes a GitLab CI configuration that runs the Single Machine Performance (SMP) regression detector so that it runs on the head of the PR (as the comparison variant) and the most recent commit of `main` (as a baseline variant) that satisfies both the following two criteria:

* is accessible from the commit history of PR's HEAD
* has a container image in SMP's ECR (which mirrors parts of Agent CI's ECR)

To compute the SHA of the baseline variant, the regression detector job configuration starts at the merge base of the PR with `main` and queries SMP's ECR to see if an image exists for that commit. If such an image does not exist, the ECR query repeats on the first parent of the commit queried in the previous iteration. The iteration terminates when one of the following three criteria are satisfied (in decreasing order of precedence):

* an image exists for the commit
* the commit has no ancestors in `main`
* the CI job times out

If this loop terminates due to the second or third criteria, the regression detector will not be run on the PR. This situation can occur due to an age-based image expiration policy in ECR if the merge base commit is sufficiently old. In that scenario, rebasing the PR onto a sufficiently recent commit in `main` is recommended.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

We'd like to run the regression detector to compare a PR with its merge base because that comparison is easier to reason about than the current regression detector configuration. The current regression detector configuration compares a PR with the latest version of `main`, which may contain commits that are not in the PR.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

Ideally, to account for interactions resulting from merging the PR into `main`, we would compare that hypothetical merge commit to its parent from `main`. The approach proposed in this PR neglects those interactions. Both the execution time of that "ideal" approach and the developer effort required to implement it are prohibitive, which motivates the approach taken in this PR.

We believe the risks in neglecting these merge interactions can be mitigated by running a variation of the same analysis nightly that compares successive commits in `main` after the fact, if necessary. To limit review burden and development effort, we consider implementing that mitigation strategy out of scope of this PR; we mention it here for completeness.

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

Using the merge base commit as the baseline variant costs additional developer effort and slightly increases maintenance burden and complexity. We believe this cost is worthwhile because this change will make regression detector results easier to explain.

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Once complete, the changes in this PR can be tested by running the Agent CI pipeline and examining the regression detector output.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

goxberry committed:
regression detector: add compute merge base job

We need to compute the merge base of a non-`main` branch with respect
to `main` in order to establish the commit SHA of a baseline variant
in regression detection. This commit adds a job to compute that merge
base, echo the result to a file (sans newline), and then upload that
result to the Single-Machine Performance S3 bucket for Agent team.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: note build step we may need

The existing Single-Machine Performance (SMP) regression detector
setup doesn't build additional containers -- it just publishes
containers already built in Agent CI to SMP's ECR for Agent
images. I've written the new "merge-base" container assuming we can
continue with that strategy, but if we can't continue with that
strategy for some reason, then this commit includes a bunch of
comments sketching out a backup plan that would build the container we
would need and publish it to SMP's ECR for Agent images.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detectr: sketch functional test tweaks

As in the spirit of the previous commit, this commit adds some
comments on how to transition the existing setup in
`functional_test/regression_detector.yml` from using the latest commit
on `main` to using a "merge base" baseline SHA.

There are a few parts of this commit that are actually functional, but
shouldn't functionally alter the existing regression detector output
-- it still uses a "latest `main`" baseline SHA -- but it does
introduce the idea of getting the new baseline SHA from an artifact in
a previous stage, rather than uploading it to S3 (although I also
upload the merge base baseline SHA to S3 as well, as a contingency
plan).

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
smp, docker_linux.yml: use literal for image

I thought using a `!reference` tag would work for specifying an image
for the merge-base-computing job, but that mental model may be wrong,
so this commit replaces that tag with the literal it's supposed to
(de)reference.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
smp, docker_linux.yml: comment out merge base job

The merge base job isn't running because no runners are configured to
run it. This issue could be one that I can't resolve, so it could be
that I need to get permissions to run it. For now, I'll comment out
this job and instead try and get the information necessary to compute
a baseline as part of the regression detector job, but without
otherwise altering the regression detector behavior.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression_detector.yml: compute merge base

As a test of whether I can compute the merge base using a `git fetch`
command, this commit adds that command (and other supporting commands)
to the regression detector job to see if these commands will work.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
smp, docker_linux.yml: add container build notes

I've managed to figure out how to get the most important case to work
-- the one in which the base branch of the pull request (i.e., in
GitLab terms, the target branch of the merge request) is
`main`. Having managed to get this case to work, this commit updates
my implementation notes to reflect how to implement the case when the
base branch of the pull request *is not* `main`.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: set baseline to merge base

It looks like creating a separate job (at least in
`.gitlab/container_build/docker_linux.yml`) will not work because
GitLab won't run it in a runner. This behavior may be a configuration
setting for security reasons, and may require repo admin privileges to
change. Instead, this commit implements setting the baseline SHA to
the merge base, along with some commands to abort the regression
detector job if the base branch of the pull request (target branch of
the merge request) is not `main`, and copies the merge base SHA to S3
for debugging/auditing purposes.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression_detector.yml: remove obsolete comments

Now that I'm pretty sure I've figured out how to change the baseline
SHA to the merge base of a pull request, this commit deletes most of
the commented-out lines I introduced into
`.gitlab/functional_test/regression_detector.yml` as notes to
myself. These comments are no longer necessary for record-keeping.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry commented on .gitlab/container_build/docker_linux.yml:
```diff
@@ -119,6 +119,49 @@ docker_build_agent7:
+#     - aws configure set aws_secret_access_key $(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.single-machine-performance-bot-access-key --with-decryption --query "Parameter.Value" --out text) --profile single-machine-performance
+#     - aws configure set region us-west-2 --profile single-machine-performance
+#     - aws s3 cp --profile single-machine-performance --only-show-errors "${CI_COMMIT_SHA}-baseline_sha" "s3://${SMP_AGENT_TEAM_ID}-smp-artifacts/information/"
+#     # NOTE(geoffrey.oxberry@datadoghq.com): For the purposes of standing up a
+#     # minimum viable product, assume a container has already been built for the
+#     # merge base -- this assumption is actually true if the target branch of the
+#     # pull request is `main`. If not, we need to (sort of) extend
+#     # .docker_build_on_job_definition, .docker_build_job_definition_amd64, and
+#     # docker_build_agent7. Using the `extends:` keyword won't actually work
+#     # because it'll reverse deep merge settings, so instead this hypothetical
+#     # build task could use `!reference` tags liberally, except for the `script`
+#     # keyword, which would be some suitably modified version of the contents of
+#     # `.docker_build_on_job_definition:script`. As far as I can tell, we only need
+#     # to build a container if the target branch is *not* `main`; it may make sense
+#     # to simply exclude that case.
+#     #
+#     # If we need to build that container, we probably also need to publish it,
+#     # which could be done by suitably modifying a *renamed copy* of
+#     # `single_machine_performance-amd64-a7` (because we still need the existing
+#     # version of that task). If we need to inject a different version of the
+#     # shell variables that relies in some way on outputs from another SMP job in
+#     # Agent CI, we can use the `before_script` hook to `export` the variable
+#     # with the value we need.
```
If we were to restrict the base branch of a pull request to `main`, would that suffice? If so, I think this pull request, as of https://github.com/DataDog/datadog-agent/pull/17449/commits/8dd9f11a9ad8417b3b5622671f95f232f4f7d562, is pretty close to the desired end state of this PR. If not, I'll have some additional questions on building and publishing containers.

bkabrda replied:
I think we can restrict to `main` and "stable branches", which are `7.45.x`, `7.44.x` etc.

goxberry replied:
Makes sense to me. @scottopell has pointed out some [recent regression detector results](https://github.com/datadog/datadog-agent/pull/17160#issuecomment-1549872166) that would benefit from setting the baseline to a merge base with `main`, so I'd like to limit this PR to that case, then augment that case with an implementation for other stable branches in a subsequent PR. For other stable branches, we'd need to figure out whether a merge base with a branch like `7.45.x` would require building and publishing a container for that merge base. I'm hoping the other operations needed are straightforward (querying the GitHub API for the PR's base branch, checking that this base branch is a stable branch).

goxberry committed:
docker_linux.yml: correct line comment

This commit corrects a line comment in which I write "Use this line if
comparing `main` against itself makes sense" when I should have
written "Use this line if comparing `main` against itself does not
make sense".

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression_detector.yml: make if block a one-liner

GitLab apparently doesn't echo multiline commands by default, so this
commit rewrites this multiline `if` block as a one line command for
debugging purposes.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression_detector.yml: remove backticks

I'm pretty sure that the "command `main` not found" message I saw was
because the shell was likely interpreting those backticks as running a
command. This commit replaces the backitcks with single quotes to
avoid having the shell attempt to run a command called `main`.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression_detector.yml: disable base branch check

This commit temporarily disables the check for the base branch because
it doesn't quite seem to work at the moment, and it adds enough
debugging output so I can get some idea of why that check does not
work.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression_detector.yml: add artifacts + cleanup

This commit cleans up the last bits of debugging comments and code in
`.gitlab/functional_test/regression_detector.yml`, plus it adds the
reporting outputs as artifacts to provide additional diagnostic output
for debugging purposes.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

bkabrda reviewed (commented):
Just two small notes for now, we'll need to discuss this in more detail.

goxberry commented on .gitlab/container_build/docker_linux.yml:
```diff
@@ -119,6 +119,49 @@ docker_build_agent7:
+#   rules:
+#     - !reference [.on_a7]
+#     # NOTE(geoffrey.oxberry@datadoghq.com): Assume for now it's sensible to compare main against itself.
+#     # - !reference [.if_not_main_branch]    # Use this line if comparing main against itself does not make sense.
```
Theoretically, comparing `main` against itself could be used for sanity-checking, but it isn't necessary here.

goxberry committed:
docker_linux.yml: remove comment cruft

This commit removes a large comment block I put in
`.gitlab/container_build/docker_linux.yml` because I no longer think
it's a good idea to compute the merge base commit in a job separate
from the regression detector job.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: compute baseline via loop

The previous regression detector baseline SHA computation assumed that
an image exists in ECR for the commit returned by `$(git commit
merge-base HEAD main)`, *i.e.*, the merge base with `main`. However,
the container associated with that commit may not exist because that
container may have failed to build, or may have failed to upload to
ECR, so we need to check whether that container exists in ECR. If that
container exists, then the merge base with `main` is the baseline
SHA. If not, then we must iterate over predecessor commits in `main`
until we find a commit in `main` for which a container exists in
ECR. The first commit we find in this loop becomes the baseline SHA
for the regression detector.

This commit implements that check, along with a loop that iterates
over predecessor commits, if necessary, as described above. The
implementation is currently a rather verbose one-liner because
single-line statements are easier to debug in GitLab CI. Once this
implementation succeeds, a subsequent commit will clean up the
implementation by changing it to a multi-line statement, after which
this branch will be ready for review.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: reduce stderr noise

The initial implementation of the container existence check turns out
to be pretty noisy when a container doesn't exist because `awscli`
will output a bunch of error information on failure. While this
information is helpful for debugging purposes, it will be annoying in
CI, so this commit redirects the `stderr` of that command to
`/dev/null` to reduce noise in the CI output of the regression
detector CI job.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: fix copy before write error

The regression detector job in the Agent CI pipeline currently fails
because it attempts to copy a file that doesn't exist to S3. This
commit fixes that error by moving the copy command to a point after
the file is generated.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: tweak ecr query for debugging

For some reason, the `aws ecr describe-images` command is not finding
images that I know exist. To aid in debugging at the expense of noise,
this commit removes redirection of `stderr` to `/dev/null`, adds the
`--registry-id` flag to be more explicit about ECR repo location, and
also moves the `--profile` flag and its argument to a more readable
location.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: remove "latest main" job

The baseline-SHA-computing steps in the
`single-machine-performance-regression_detector` job make computing
the commit of "latest `main`" unnecessary, so this commit deletes the
job that does that computation.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry committed:
regression detector: remove loop cleanup comment

Due to falling behind on Agent performance investigations, I don't
think I'm going to get to clean up the long, one line loop statement,
so this commit deletes that aspirational comment -- it can be cleaned
up in a subsequent pull request.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry commented on .gitlab/container_build/docker_linux.yml:
```diff
@@ -119,6 +119,49 @@ docker_build_agent7:
+#     - aws configure set aws_access_key_id $(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.single-machine-performance-bot-access-key-id --with-decryption --query "Parameter.Value" --out text) --profile single-machine-performance
+#     - aws configure set aws_secret_access_key $(aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.single-machine-performance-bot-access-key --with-decryption --query "Parameter.Value" --out text) --profile single-machine-performance
+#     - aws configure set region us-west-2 --profile single-machine-performance
+#     - aws s3 cp --profile single-machine-performance --only-show-errors "${CI_COMMIT_SHA}-baseline_sha" "s3://${SMP_AGENT_TEAM_ID}-smp-artifacts/information/"
```
For the sake of expedience, I decided to move the baseline SHA computation into the regression detector job itself.

goxberry committed:
regression detector: pin pr-commenter version

This commit attempts to triage the `pr-commenter` errors by pinning
the `pr-commenter` version in a fashion similar to that used by
other Datadog repositories (e.g., `dogweb`).

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry commented on .gitlab/functional_test/regression_detector.yml:
```diff
@@ -41,11 +52,15 @@ single-machine-performance-regression_detector:
+    - BASELINE_SHA="${SMP_MERGE_BASE}"  # NOTE(geoffrey.oxberry@datadoghq.com): cat the echoed file once computing merge commit is separate job
+    - echo "Computing baseline..."
+    - echo "Checking if image exists for commit ${BASELINE_SHA}..."
+    - while [[ ! $(aws ecr describe-images --profile single-machine-performance --registry-id "${SMP_ACCOUNT_ID}" --repository-name "${SMP_AGENT_TEAM_ID}-agent" --image-ids imageTag="${BASELINE_SHA}-7-amd64") ]]; do echo "No image exists for ${BASELINE_SHA} - checking predecessor of ${BASELINE_SHA} next"; BASELINE_SHA=$(git rev-parse ${BASELINE_SHA}^); echo "Checking if image exists for commit ${ ...
```
@bkabrda I've added a loop that will iterate backwards over commits on `main` until it finds a commit with an image in SMP's ECR, starting with the merge base commit. Adding this loop should make the regression detector job match @KSerrania's "option 2" proposal for the regression detector baseline.

goxberry committed:
regression detector: delete more stale comments

For now, I don't plan to move the baseline SHA computation into a
separate job, although that may happen later. Given this change of
plans, this commit removes the stale comment regarding refactoring the
baseline SHA computation to a separate job.

Signed-off-by: Geoffrey M. Oxberry <geoffrey.oxberry@datadoghq.com>

goxberry commented on .gitlab/functional_test/regression_detector.yml:
```diff
@@ -103,7 +117,8 @@ single-machine-performance-regression_detector:
     - echo ""
-    - dd-package --bucket binaries.ddbuild.io --package devtools/pr-commenter
+    # Pinned version of pr-commenter taken from https://github.com/DataDog/dogweb/blob/b2a891c2a2186b8fecd27ca9b13ebabe5f289a4a/tasks/gitlab/k8s-diff-helper.sh#L22
+    - dd-package --bucket binaries.ddbuild.io --package devtools/pr-commenter --distribution "20.04" --version "14692290-a6440fd1"
```
Pinning the `pr-commenter` version seems to work for the time being, but it's likely the `pr-commenter` version will need to be updated because the CI Interfaces team has been adding telemetry to that service. [Recent versions of `pr-commenter`](https://github.com/DataDog/devtools/tree/fbb16979b1c97ed1a55e60b8a9e3acecc2728ddf/deb/pr-commenter) require a GitHub App key. For provenance purposes, it could make more sense for Agent to own that key.

goxberry replied:
re: this specific line, without this change, the regression detector is definitely broken due to upstream changes in `pr-commenter`, so I'd be inclined to defer more `pr-commenter` changes to a later PR.

goxberry commented:
Updated pull request description to reflect current status.

bkabrda reviewed (approved):
LGTM, I think this puts us in a good spot with running the regression detector for PRs for `main` branch.
